app/response_processor.py

```diff
@@ -46,105 +52,129 @@ def service_name(self, url=None):
             elif 'validate' in parts:
                 return 'SDX-VALIDATE'
         except AttributeError as e:
-            self.logger.error(e)
+            self.logger.error("No valid service name", exception=e)
 
-    def process(self, encrypted_survey):
-        # decrypt
+    def _process(self, encrypted_survey):
         decrypted_json = self.decrypt_survey(encrypted_survey)
 
-        metadata = decrypted_json['metadata']
-        self.logger = self.logger.bind(user_id=metadata['user_id'], ru_ref=metadata['ru_ref'])
+        metadata = decrypted_json.get('metadata', {})
+        self.logger = self.logger.bind(user_id=metadata.get('user_id'),
+                                       ru_ref=metadata.get('ru_ref'))
 
         if not self.tx_id:
             self.tx_id = decrypted_json.get('tx_id')
         elif self.tx_id != decrypted_json.get('tx_id'):
-            raise BadMessageError
+            logger.info('tx_ids from decrypted_json and message header do not match.' +
+                        ' Rejecting message',
+                        decrypted_tx_id=decrypted_json.get('tx_id'),
+                        message_tx_id=self.tx_id)
+            raise QuarantinableError
 
         self.logger = self.logger.bind(tx_id=self.tx_id)
 
         try:
             self.validate_survey(decrypted_json)
-        except BadMessageError:
+        except ClientError:
             # If the validation fails, the message is to be marked "invalid"
             # and then stored. We don't then want to stop processing at this point.
+
             decrypted_json['invalid'] = True
-            self.logger.info("Invalid survey data, skipping receipting", tx_id=self.tx_id)
+            self.logger.info("Invalid survey data, skipping receipting")
+
+        if decrypted_json.get("survey_id") != "feedback" and decrypted_json.get('invalid') is not True:
+            self.logger.info("Receipting survey")
+            self.send_receipt(decrypted_json)
         else:
-            # only send the receipt is the json is valid
-            if decrypted_json.get("survey_id") != "feedback":
-                self.logger.info("Receipting survey", tx_id=self.tx_id)
-                self.send_receipt(decrypted_json)
-            else:
-                self.logger.info("Feedback survey, skipping receipting", tx_id=self.tx_id)
-        finally:
-            # store the survey regardless
-            self.store_survey(decrypted_json)
+            self.logger.info("Feedback survey, skipping receipting")
```

Comment:
If the survey is invalid then this will log out "Feedback survey, skipping receipting" even if it isn't the feedback survey. I think the logic was clearer before, please change back